File: python/src/day14/main.py
from src.common.file_utils import get_path, read_lines
import re

BITS = 36


def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    memory = {}

    for line in lines:
        splitted = line.split("=")
        key, val = splitted[0].strip(), splitted[1].strip()
        
        if key == "mask":
            mask = val
        else:
            address = int(re.search(r'mem\[(\d+)]', key).groups()[0])
            val = list('{:036b}'.format(int(val)))
            for idx, c in enumerate(mask):
                if c != "X" and c != val[idx]:
                    val[idx] = c
            memory[address] = int("".join(val), 2)
    
    return sum(memory.values())


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    memory = {}
    
    for line in lines:
        splitted = line.split("=")
        key, val = splitted[0].strip(), splitted[1].strip()
       
        if key == "mask":
            mask = val.strip()
        else:
            address = int(re.search(r'mem\[(\d+)]', key).groups()[0])
            address_base = list('{:036b}'.format(address))
            floating = 0
            # replace with 1's, mark x's with {} for formatting
            for i in range(BITS):
                if mask[i] == "X":
                    address_base[i] = "{}"
                    floating += 1
                elif mask[i] == "1":
                    address_base[i] = "1"
            format_address = "".join(address_base)
            if floating > 0:
                bin_format = "0{}b".format(floating)
                # insert each binary number up to 2**floating into positions of X's
                for i in range(2 ** floating):
                    mem_address = format_address.format(*format(i, bin_format))
                    memory[int(mem_address, 2)] = int(val)
            else:
                memory[int("".join(format_address), 2)] = int(val)

    return sum(memory.values())


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day14/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var (
	bits = 36
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	lines := input.ReadLines(filename)
	memory := make(map[int]int)
	mask := ""

	for _, line := range lines {
		splitted := strings.Split(line, "=")
		key, val := strings.TrimSpace(splitted[0]), strings.TrimSpace(splitted[1])

		if key == "mask" {
			mask = val
			continue
		}

		r := regexp.MustCompile(`mem\[(\d+)]`)
		address := toInt(r.FindStringSubmatch(key)[1])

		binaryVal := parseBinaryStringFromDecimalString(val)
		paddedBinaryVal := leftPadString(binaryVal, bits)
		binarySlice := strings.Split(paddedBinaryVal, "")

		for idx, c := range mask {
			if string(c) != "X" {
				binarySlice[idx] = string(c)
			}
		}

		memory[address] = parseDecimalFromBinarySlice(binarySlice)
	}

	return sumVals(memory)
}

func Part2(filename string) int {
	return 12
}

func leftPadString(s string, n int) string {
	return fmt.Sprintf("%0*s", n, s)
}

func parseBinaryStringFromDecimalString(i string) string {
	decimal, _ := strconv.ParseInt(i, 10, 64)
	return strconv.FormatInt(decimal, 2)
}

func parseDecimalFromBinarySlice(b []string) int {
	i, _ := strconv.ParseInt(strings.Join(b, ""), 2, 64)
	return int(i)
}

func sumVals(m map[int]int) int {
	s := 0
	for _, val := range m {
		s += val
	}
	return s
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}